File: 10-GRAPHS/Python/max-area-of-island.py
from typing import List

# https://leetcode.com/problems/max-area-of-island/
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        if len(grid) == 0:
            return 0

        rows, cols = len(grid), len(grid[0])

        def dfs(r: int, c: int) -> int:
            if min(r, c) < 0 or r >= rows or c >= cols or grid[r][c] == 0:
                return 0

            grid[r][c] = 0

            return 1 + dfs(r - 1, c) + dfs(r + 1, c) + dfs(r, c - 1) + dfs(r, c + 1)

        result = 0
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1:
                    result = max(result, dfs(r, c))

        return result

File: 10-GRAPHS/PHP/max-area-of-island.php
<?php

// https://leetcode.com/problems/max-area-of-island/
class Solution
{
    private array $grid = [];
    private int $rows = 0;
    private int $cols = 0;

    /**
     * @param int[][] $grid
     * @return int
     */
    function maxAreaOfIsland(array $grid): int
    {
        if (count($grid) == 0) {
            return 0;
        }

        $this->grid = $grid;
        $this->rows = count($grid);
        $this->cols = count($grid[0]);

        $result = 0;
        for ($r = 0; $r < $this->rows; $r++) {
            for ($c = 0; $c < $this->cols; $c++) {
                if ($this->grid[$r][$c] == 1) {
                    $result = max($result, $this->dfs($r, $c));
                }
            }
        }
        return $result;
    }

    private function dfs(int $r, int $c): int
    {
        if ($this->isOutOfBound($r, $c) || !$this->grid[$r][$c]) {
            return 0;
        }

        $this->grid[$r][$c] = 0;

        return 1
            + $this->dfs($r - 1, $c)
            + $this->dfs($r + 1, $c)
            + $this->dfs($r, $c - 1)
            + $this->dfs($r, $c + 1);
    }

    private function isOutOfBound(int $r, int $c): bool
    {
        return min($r, $c) < 0 || $r >= $this->rows || $c >= $this->cols;
    }

}